### ISSUE TITLE
manylatents-omics has no scRNA preprocessing module — filter/normalize/transform/doublets/HVG, as pure functions returning masks

### ISSUE THREAD

**Issue by cmvcordova** [latent-reasoning-works/manylatents#297]
**Related:** #292 (HVG + granger wiring) — **different HVG, see §5 before you start.**
**Consumer:** `geomancer`'s `prep` step group (landed: latent-reasoning-works/geomancer @ 328c96e).

## The ask

A scRNA preprocessing module in manylatents-omics: `filter_cells`, `filter_genes`, `filter_mito`,
`normalize`, `transform`, `detect_doublets`, `highly_variable_genes`.

## Why here and not in the consumer

geomancer is the product layer and implements no compute — its `CLAUDE.md` is explicit that DR
algorithms, metrics and training loops belong in `manylatents`/`geomancy`. It has a `prep` step
group ready to dispatch to these and a recipe format that declares their hyperparameters, but
nowhere honest to put the functions themselves.

## What exists today — measured, not assumed

Nothing. Grepped across the installed `manylatents` at the `main` geomancer pins:

- `normalize_total`, `filter_cells`, `filter_genes`, `highly_variable_genes`, `scrublet` →
  **zero hits** outside `singlecell/algorithms/_scgpt_vendor/preprocess.py` (a vendored scGPT
  copy, not an API) and `popgen/utils/data.py` (`hwe_normalize`, genotypes, not scRNA).
- The only preprocessing-shaped thing in `singlecell/` is
  `scgpt_encoder._preprocess_adata` — private, inside one model.

`manylatents/singlecell/` already has `algorithms/`, `analysis/`, `configs/`, `data/` and
`sampling.py`, so `singlecell/preprocessing.py` is the obvious home.

## The contract the consumer needs

Each operation is a **pure function**: it computes, it does not mutate, and a filter returns a
**boolean mask** rather than a filtered object.

```python
def filter_cells(counts, *, min_genes=200) -> np.ndarray:   # bool, len == n_cells
def filter_genes(counts, *, min_cells=3)  -> np.ndarray:    # bool, len == n_genes
def filter_mito(counts, gene_names, *, max_pct=5.0, prefix="MT-") -> np.ndarray
def detect_doublets(counts, *, threshold=None) -> np.ndarray  # bool: is_doublet
def normalize(counts, *, target_sum=1e4)  -> np.ndarray     # a NEW matrix
def transform(X, *, method="log1p")       -> np.ndarray     # log1p | sqrt
def highly_variable_genes(counts, *, n_top_genes=2000, flavor="seurat_v3") -> np.ndarray
```

Three properties, and each of them is load-bearing downstream rather than stylistic:

1. **A mask, not a filtered object.** The consumer keeps the loaded matrix immutable and carries
   a selection into it, so that "which cells did this drop, and why" stays answerable at any
   later point. A function that returns a filtered copy destroys that and forces a diff.
2. **No mutation of the input.** `sc.pp.normalize_total` writes **in place**, and the consumer's
   matrix is the user's own `.X` (an 18.3 MB CSR on pbmc3k, 353.6 MB densified). Copy before
   any in-place scanpy call.
3. **Sparse in, sparse out where possible.** `counts > 0` on a CSR touches stored entries only.
   Densifying pbmc3k to answer "how many genes per cell" costs 353.6 MB for a column sum.

`detect_doublets` returning `is_doublet` rather than `keep` is deliberate: the consumer's default
is to **flag without removing**, and inverting a mask is cheaper than recovering a discarded call.

## §5 — the HVG in #292 is NOT the HVG here

This is the one thing worth reading twice, because the two share a name and nothing else.

| | #292's HVG | this issue's HVG |
|---|---|---|
| input | genes × trajectories (`[T, n_genes]`, time-major) | cells × genes count matrix |
| source | decoded from a trained flow over a time grid | the raw `.h5ad` as loaded |
| selects on | variability **across pseudotime** | variability **across cells** |
| feeds | `granger_grn` | the embedding, and every downstream step |
| when | after training | before anything |

The backlog wording for #292 is *"highly variable genes based of **gene trends**"* — trends, i.e.
trajectories. That is a genuinely different statistic from `sc.pp.highly_variable_genes` over a
count matrix, and implementing one will not give you the other. If they end up in the same
module they need different names; `hvg_over_time` vs `hvg` would do.

## Non-goals

- **No orchestration.** Order, defaults and which operations run are the consumer's recipe, not
  this module's business. Do not add a `preprocess()` that chains them.
- **No AnnData requirement in the signatures.** Take arrays and gene names. The consumer holds
  an AnnData and can pass `.X` and `.var_names`; a signature that demands the object makes the
  module untestable on a synthetic matrix.
- **No config/Hydra wiring needed** for the consumer's sake — it declares hyperparameters in its
  own recipe format. Add Hydra configs if manylatents wants them for its own runs.

## What happens on the consumer side

geomancer's `prep` executors become thin dispatch with no scanpy import, and its recipes change
`via: scanpy` → `via: manylatents`. Its step group, precondition calculus and record already
handle narrowing (a filter clears every derived array, because an embedding fitted over cells
that are now gone is not an embedding of the ones that remain). Nothing else has to move.

Happy to take contract questions — the mask/immutability constraints come from a design already
landed on the consumer side, so if any of them are awkward here it is worth saying before you
build against them.

**Comment by cmvcordova** [comment]
Moved to latent-reasoning-works/manylatents-omics#60 — this asks for a module under `manylatents/singlecell/`, which is shipped by the separate **manylatents-omics** repo (it extends the `manylatents` namespace with `singlecell`, `dogma` and `popgen`), not by core manylatents, which has no `singlecell/` directory at all. Same body and same assignees over there, with the `#292` references fully qualified and two path claims corrected against the omics tree; closing this one as not planned.
